src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+/// variable to concrete type.
+/// \param generic_type The java_generic_typet that should be modified and
+/// instantiated.
+/// \param concrete_types The mapping of type variables to concrete types.
+const void generate_java_generic_methodt::instantiate_java_generic_type(
+  java_generic_typet &generic_type,
+  const generate_java_generic_methodt::type_variable_instantiation_mapt
+    &concrete_types) const
+{
+  INVARIANT(generic_type.id()==ID_pointer,
+            "All generic parameters should be pointers in java");
+  INVARIANT(generic_type.subtype().id()==ID_symbol,
+            "All generic parameters should point to symbols");
+
+  // Replace the generic parameters
+  for(java_generic_parametert &generic_parameter :
+    to_java_generic_type(generic_type).generic_type_variables())
+  {
+    if(!is_java_generic_inst_parameter(generic_parameter))
+    {
+      const symbol_typet &generic_type_variable=generic_parameter
+        .type_variable();
+      const auto &instantiation_type=concrete_types.find
+        (generic_type_variable);
+      if(instantiation_type==concrete_types.end())
+      {
+        // If we ever want/need to support partially instantiated types,
+        // we'll probably want to just return here.
+        throw "no concrete type supplied for generic type parameter";
+      }
+      generic_parameter.subtype()=instantiation_type->second;
+      generic_parameter.set(ID_C_java_generic_inst_parameter, true);
+    }
+  }
+  // FIXME: At this point we have a concrete types substituted into the
+  // FIXME: generic type variables, but it's still a generic type.
+  // FIXME: I think once the rest of the generics stuff is committed,
+  // FIXME: particularly, once we have specialization of classes working,
+  // FIXME: we'll want to then trigger and/or replace the whole generic type
+  // FIXME: expression here with the concrete type.
+}
+
+/// Decorate an identifier on a given expression if the exisiting identifier
+/// matches a given pattern.
+/// \param expr The expression whos identifier should be decorated
+/// \param identifier The particular identifier that should be modified
+/// \param pattern The pattern to match and be decorated
+/// \param decoration The string to use as decoration
+///
+/// An example of the use of this function might be:
+///    decorate_identifier(
+///      symbol,
+///      ID_identifier,
+///      "java::array[boolean].clone:()Ljava/lang/Object;"
+///      "<decoration>")
+/// which when given a 'symbol' expression who's 'ID_identifier' looks like:
+///   'java::array[boolean].clone:()Ljava/lang/Object;::this' would replace
+/// the identifier with:
+///   'java::array[boolean].clone:()Ljava/lang/Object;<decoration>::this'
+/// If the given 'expr' parameter does not already contain an identifier
+/// whos value begins with the given 'pattern' string, then no change will
+/// be made.
+void generate_java_generic_methodt::decorate_identifier(
+  irept &expr,
+  const dstringt &identifier,
```

Comment:
Nit: `irep_idt` is typedef'd to `dstringt`, but I think it is always preferable to use `irep_idt`

Reply:
Good point, will fix that up.